### DISCUSSION TITLE
Flushing, Indexing, Sealing Segments, Deletions and num_entities

### DISCUSSION THREAD

**Discussion by datenhahn** [milvus-io/milvus#19259] [Q&A and General discussion] [answered]
Hi,

I am not sure if this should be a bug report or a discussion, so I am starting here for now.

I am setting up MIlvus for a customer in the publishing business and we plan to store a couple of millions of document vectors in milvus. Documents come in from a kafka topic in a rate of about 60 documents per Minute.

What we noticed is that while inserting documents the number of vectors in the collection in the attu Web-UI does not update. When we call the infamous "num_entities" on a collection via the python client it takes quite some while and then the numbers are accurate. It seems to take longer the longer we wait.

We tested this with Milvus Version v2.1.2 (and a couple of versions before) and Kafka as Message-Layer and S3 as persistence layer.

Regarding some other ticket I opened we got this additional info:

https://github.com/milvus-io/milvus/issues/18565#issuecomment-1214794545

_We seal a segment whenever called num_entities now. With this in mind, I suggest don't call num_entities after every insert/delete. In order to get numOfRows, you can use get_query_segment_info_

I took more info from this page: https://milvus.io/blog/2022-2-21-compact.md
_Another issue is that a segment with less than 1,024 rows is not indexed and only supports brute-force search to process queries. Small segments caused by auto-flush or user-invoked flush might hamper query efficiency._

And finally this page about deletion: https://milvus.io/blog/2022-02-07-how-milvus-deletes-streaming-data-in-distributed-cluster.md

So I have the following questions:

- If we insert a vector. From which moment on is the vector available for search? (see below for the search function call, is this a "brute-force" call? What is a brute force call?)
- What causes a segment to get sealed/flushed to S3 (or other persistence), it seems without calling any function which triggers the sealing (e.g. num_entities) the sealing/flushing does not happen (there many thousand vectors inserted without a flush happening)
- Are there config parameters regarding flushing/sealing of segments which can be configured and should be tuned (according to what goals?)
- How do restarts of any nodes affect deletions? As far as I understood until compaction happens and flushes the deletes into the sealed segments the datanode/querynode need to consume the DML and/or DELTA topics and filter out deleted data on-the-fly. Do the datanodes/querynodes automatically reset their consumergroups/kafka offsets when they get restarted and reprocess all deletes or do they have some other means of persistence of their current runtime state (S3? Ephemeral Volumes?)
- How long should the retention of the DML/DELTA kafka topics be? Should it be minutes, days or weeks? Is the data still relevant once processed by all nodes? What happens if Kafka Data is reset? Is everything still functional (Milvus Cluster then is loaded from S3 / etcd state, potentially only loosing some inserts/deletes) or is Kafka Data to be treated as part of the consistent state? (e.g. for backups).
- I find it a bit dangerous to have something like a "num_entities" function, which has no hint in the documentation in any way that it can hurt the cluster functionality up to making it completely unusable. A more fitting name would be "force_flush_all_segments" with a documentation comment something like "WARNING: calling this function forces the sealing of any open segments, potentially causing lots of very small segments which cannot be indexed. Only call if you know what you are doing!" .
- Are there more functions which can hurt performance of functionality of milvus?
- Are there setups which make milvus unstable (e.g. we used lots of partitions with very little vectors and this resulted in golang panics, reducing the number of partitions and packing more vectors into them worked for us to get a more stable experience)
- Is it ok to randomly delete all over the place? In our scenario the document data often changes and we want to update vectors, so we delete the vector and insert a new vector with the same id (so actually we just would like something like UPSERTs). Any vector is up for frequent deletion and reinsertion. Is there a more clever way to solve this?

This is our search call:
```
    query_results = collection.search(
        data=list_of_query_vectors,
        anns_field="embeddings",
        param=search_params,
        partition_names=search_partitions,
        limit=request.num_similar_documents,
        consistency_level="Strong"

    )
```

**Comment by yhmo** [comment, marked as answer]
- **If we insert a vector. From which moment on is the vector available for search? (see below for the search function call, is this a "brute-force" call? What is a brute force call?)**

Milvus use timestamp to ensure data consistency accross datanodes/querynodes. As we know, for insert operation , the proxy will split the data to multiple parts according to collection's shard number, and send these parts into different data channel, then each datanode/querynode will subscribe a part of data from a data channel. Then how to make data consitency accross datanodes/querynodes? The solution is: rootcoor will send a special message to data channel in a time interval(200ms), the datanode/querynode consume data from data channel, they also receive the special message from the data channel. These nodes report to rootcoord which message they have consumed, so the rootcoord knows how many operations have been consumed completely.
For example, let's say there are some insert operations Op_1/op_2/op3/... , and the rootcoord send special messages m_1/m_2/m_3/... by interval 200ms. The time line looks like:
op_1 -> op_2 -> m_1 -> op_3 -> m_2 -> op_4 -> op_5 -> op_6 -> m3 -> op_7

The querynode_1 reports it consume to m_2, so the rootcoord knows the querynode_1 consumed the data of op_1, op_2 and op_3
The querynode_2 reports it consume to m_3, so the rootcoord knows the querynode_2 consumed the data of op_1, op_2, op_3, op_4, op_5, op_6

The two querynodes are on different position in the timeline, querynode_1 is on m_2, querynode_2 is on m_3. So, the position we can ensure data consistency is m_2, because the op_4/op_5/op_6 are invisible to querynode_1.

User call search() after op_7, then which operations are visible to the user? It depends on the "consistency_level" parameter.
If consistency_level="Strong", milvus will wait until the op_7 is both consumed by querynode_1 and querynode_2.
If consistency_level="Bound", and if the current consume point m_2 is close to current timestamp(smaller than 5 seconds), milvus can execute search instantly. Otherwise, milvus will wait next consume point until the latest consume point is close to current timestamp.
If consistency_level="Eventually", milvus will execute search instantly, by using current consume point m_2 as search scope.

The upper machinery is nothing related with "brute-force search". "brute-force search" is calculate distance for all the vectors in a segment. This happens on segment whcih is not finish index or not flushed.


- **What causes a segment to get sealed/flushed to S3 (or other persistence), it seems without calling any function which triggers the sealing (e.g. num_entities) the sealing/flushing does not happen (there many thousand vectors inserted without a flush happening)**

As we know, querynode/datanode consume data from data channel. For example, an insert action inserts 10000 entities into a collection whose shard number is 2. Let's say 5100 entities go to data_channel_1, and 4900 entities go to data_channel_2. The querynode_1 receive 5100 entities, querynode_2 receive 4900 entities. If num_entities is not called, the entities will stay in querynode's memory as a "growing" segment, this kind of segment has no index. Milvus will automatically trigger a flush action if the segment's size reaches 16MB. The datanode_1 also consume the 5100 entities, datanode_2 also consume the 4900 entities, but these entities are only prepare for flush, not for search. Once a flush action is triggered (by manually or automatically), the datanodes will flush these entities to persist a segment into storage. And the "growing" segment in querynode will become "sealed". Then the indexnode take this persised segment to build index, after index is finish, the "sealed" segment in querynode will be replaced by the index. Before the index is finish, "brute-force search" will be applied on the "sealed" segment in querynode.

In v2.1, call num_entity method will internal call a flush() method to cause a segment to flushed and sealed. But in v2.2, this behavior will change, the num_entity won't cause any flush action.


- **Are there config parameters regarding flushing/sealing of segments which can be configured and should be tuned (according to what goals?)**

In milvus.yml there is an item can control the segment size threshold to trigger a auto-flush action:
```
dataNode:
  flush:
    # Max buffer size to flush for a single segment.
    insertBufSize: 16777216 # Bytes, 16 MB
```

If the insertBufSize value is large, the "growing" segment in querynode is large, "brute-force search" on large segment will slow down search performance.
If the insertBufSize value is small, the "growing" segment in querynode is small. The persisted segment also is small, then small segments will be merged by compaction machinery. Lot of small segments might lead to heavy work on compaction and datanode/indexnode slow response.


- **How do restarts of any nodes affect deletions? As far as I understood until compaction happens and flushes the deletes into the sealed segments the datanode/querynode need to consume the DML and/or DELTA topics and filter out deleted data on-the-fly. Do the datanodes/querynodes automatically reset their consumergroups/kafka offsets when they get restarted and reprocess all deletes or do they have some other means of persistence of their current runtime state (S3? Ephemeral Volumes?)**

Insert/delete actions are passed by data channel, datanode subscribes the data channel. If datanode receive a deletion, it firstly use bloom filter to mapping the delted ids to correct segment, and then send the deletion into a "delta" channel(also is a pulsa/kafka channel). Querynode subscribe deletions from delta channel.
If datanode restarted, it re-consume data from the data channel, resent deletions to delta channel. Querynode receive the deletions by nothing changed since deletion idempotent.
If querynode restarted, it re-consume deletion from the delta channel to build a bitset in memory for "growing" segment.


- **How long should the retention of the DML/DELTA kafka topics be? Should it be minutes, days or weeks? Is the data still relevant once processed by all nodes? What happens if Kafka Data is reset? Is everything still functional (Milvus Cluster then is loaded from S3 / etcd state, potentially only loosing some inserts/deletes) or is Kafka Data to be treated as part of the consistent state? (e.g. for backups).**

Retention of kafka/pulsar depends on configurations of the kafka/pulsar service. The default retention is 7 days for kafka.
Ideally, if kafka/pulsa is reset, milvus ought to can work correctly. But we didn't test this case, I am not sure about it.



- **I find it a bit dangerous to have something like a "num_entities" function, which has no hint in the documentation in any way that it can hurt the cluster functionality up to making it completely unusable. A more fitting name would be "force_flush_all_segments" with a documentation comment something like "WARNING: calling this function forces the sealing of any open segments, potentially causing lots of very small segments which cannot be indexed. Only call if you know what you are doing!" .**

In v2.1, call num_entity method will internal call a flush() method to cause a segment to flushed and sealed. But in v2.2, this behavior will change, the num_entity won't cause any flush action.
The documentation didn't mention this method at all, I think this is a mistake. I will push this later.


- **Are there more functions which can hurt performance of functionality of milvus?**

A known issue is: if user call delete() to delete entities constantly, the delta channel will contains lot of deletions, if user restart the querynode, it will take very long time to restore. And search performance is also affected because querynode need to construct a huge bitset to filter entities.



- **Are there setups which make milvus unstable (e.g. we used lots of partitions with very little vectors and this resulted in golang panics, reducing the number of partitions and packing more vectors into them worked for us to get a more stable experience)**

The milvus.yml has an item for pre-alloc space of a segment in querynode:
queryNode:
```
  segcore:
    chunkRows: 1024 # The number of vectors in a chunk.
```

The default value is 1024, it means if a new "growing" segment is created in querynode, the querynode will pre-alloc 1024 rows space for it.
Let's say you have 10000 partitions, each partition has two "growing" segments in querynode. Although you only insert 2 vectors in each partition, there will be 10000 * 1024 rows are pre-allocated. That might cause OOM risk.



- **Is it ok to randomly delete all over the place? In our scenario the document data often changes and we want to update vectors, so we delete the vector and insert a new vector with the same id (so actually we just would like something like UPSERTs). Any vector is up for frequent deletion and reinsertion. Is there a more clever way to solve this?**

As the known issue I mentioned: if user call delete() to delete entities constantly, the delta channel will contains lot of deletions, if user restart the querynode, it will take very long time to restore. And search performance is also affected because querynode need to construct a huge bitset to filter entities.

Large quantities of deletions is not recommended. I have no idea about the workaround, but the team is working on redesign the deletion logic to optimize it.

**Comment by datenhahn** [comment]
Thanks a lot for the quick and complete answer, we will update our code / configs accordingly.
